src/diffusers/models/unet_2d_blocks.py

```diff
@@ -1485,7 +1516,7 @@ def forward(
             hidden_states = attn(
                 hidden_states,
                 encoder_hidden_states=encoder_hidden_states,
-                attention_mask=attention_mask,
+                attention_mask=attention_mask if encoder_hidden_states is None else encoder_attention_mask,
```

Comment:
```suggestion
                attention_mask=attention_mask if encoder_attention_mask is None else encoder_attention_mask,
```
This should be the correct way of doing it?

Reply:
> correct way of doing it

I sniffed `encoder_hidden_states` instead of `encoder_attention_mask` intentionally.  
it's not important which types of mask are available. neither of them has inherent precedence.

what matters is "are we computing self-attention or cross-attention".

if `encoder_hidden_states` is defined: we are computing cross-attention, and therefore the only mask that would be valid to consider, would be `encoder_attention_mask`.

Conversely: if `encoder_hidden_states is None`: we are computing self-attention. the only mask that would be valid to consider, would be `attention_mask`. even if an `encoder_attention_mask` is available: it would not be valid to use it, because it would have the wrong token length for the self-attention key.

Reply:
appreciate your thought process @Birch-san :) 

as is I don't think that's backwards compatible, I think in order to ease the process on getting this merged, we have to maintain compatibility with the convention of unet blocks which perform a single attention block (regardless of if that block is cross or self attention) being able to use the `attention_mask` argument if it's present.

I do also think it's a bit more straight forward for understanding the block body in this case.

Reply:
I have now removed the `encoder_attention_mask` param from `SimpleCrossAttn{Down,Up}Block2D,UNetMidBlock2DSimpleCrossAttn#forward`, on the basis that if a consumer knows they are using cross-attention: they can bind a cross-attn mask to the existing `attention_mask` param.

I am a bit worried that removing this argument, means that anybody using `get_down_block` and expecting a consistent `forward()` interface regardless of block type, may be surprised to find that they cannot pass `encoder_attention_mask` into this. I wonder if any tests do that. we'll see.

Reply:
removing the `encoder_attention_mask` param causes test failures:  
https://github.com/huggingface/diffusers/actions/runs/5019287061/jobs/8999602804?pr=2634

I've put the parameter back into the forward interface, but have added a comment to explain it's only there to maintain forward interface consistency.

Reply:
adding a dead param just to maintain forward interface compatibliity felt weird to me. so I've slightly backtracked this.

I'm now using this pattern in the "simple" 2D blocks, to give `attention_mask` precedence (to maintain UnCLIP compatibility), but to use the newer convention (use cross-attn mask when we detect that we are doing cross-attn) where possible, and to recommend that UnCLIP migrate to the new `encoder_attention_mask` convention for passing its cross-attn masks.

```python
if attention_mask is None:
    # if encoder_hidden_states is defined: we are doing cross-attn, so we should use cross-attn mask.
    mask = None if encoder_hidden_states is None else encoder_attention_mask
else:
    # when attention_mask is defined: we don't even check for encoder_attention_mask.
    # this is to maintain compatibility with UnCLIP, which uses 'attention_mask' param for cross-attn masks.
    # TODO: UnCLIP should express cross-attn mask via encoder_attention_mask param instead of via attention_mask.
    #       then we can simplify this whole if/else block to:
    #         mask = attention_mask if encoder_hidden_states is None else encoder_attention_mask
    mask = attention_mask
```